sdk/samples/iot/aziot_realtek_amebaD/aziot_realtek_amebaD.ino

```diff
@@ -0,0 +1,379 @@
+  size_t mqtt_password_length;
+  rc = az_iot_hub_client_sas_get_password(
+    &hub_client,
+    sas_duration,
+    sas_base64_encoded_signed_signature,
+    AZ_SPAN_EMPTY,
+    sas_token,
+    sizeof(sas_token),
+    &sas_token_length);
+  if (az_result_failed(rc))
+  {
+    Serial.print("Could not get the password: az_result return code ");
+    Serial.println(rc);
+  }
+}
+
+static int connect_to_azure_iot_hub()
+{
+  size_t client_id_length;
+  char mqtt_client_id[128];
+  if (az_result_failed(az_iot_hub_client_get_client_id(
+          &hub_client, mqtt_client_id, sizeof(mqtt_client_id) - 1, &client_id_length)))
+  {
+    Serial.println("[ERROR] Failed getting client id");
+    return 1;
+  }
+
+  char mqtt_username[128];
+  // Get the MQTT user name used to connect to IoT Hub
+  if (az_result_failed(az_iot_hub_client_get_user_name(
+          &hub_client, mqtt_username, sizeofarray(mqtt_username), NULL)))
+  {
+    printf("[ERROR] Failed to get MQTT clientId, return code\n");
+    return 1;
+  }
+
+  Serial.print("Client ID: ");
+  Serial.println(mqtt_client_id);
+
+  Serial.print("Username: ");
+  Serial.println(mqtt_username);
+
+  while (!mqtt_client.connected())
+  {
+    Serial.print("MQTT connecting ... ");
+
+    if (mqtt_client.connect(mqtt_client_id, mqtt_username, sas_token))
+    {
+      Serial.println("connected.");
+    }
+    else
+    {
+      Serial.print("[ERROR] failed, status code =");
+      Serial.print(mqtt_client.state());
+      Serial.println(". Try again in 5 seconds.");
+      // Wait 5 seconds before retrying
+      delay(5000);
+    }
+  }
+
+  mqtt_client.subscribe(AZ_IOT_HUB_CLIENT_C2D_SUBSCRIBE_TOPIC);
+
+  return 0;
+}
+
+void establishConnection() 
+{
+  connectToWiFi();
+
+  initializeClients();
+
+  generate_sas_key();
+
+  connect_to_azure_iot_hub();
+
+  digitalWrite(LED_PIN, LOW);
+}
+
+void setup()
+{
+  pinMode(LED_PIN, OUTPUT);
+  digitalWrite(LED_PIN, HIGH);
+  createNullTerminatedRootCert();
+  establishConnection();
+}
+
+static char* get_telemetry_payload()
+{
+  az_span temp_span = az_span_create(telemetry_payload, sizeof(telemetry_payload));
+  temp_span = az_span_copy(temp_span, AZ_SPAN_FROM_STR("{ \"deviceId\": \"" IOT_CONFIG_DEVICE_ID "\", \"msgCount\": "));
+  (void)az_span_u32toa(temp_span, telemetry_send_count++, &temp_span);  
+  temp_span = az_span_copy(temp_span, AZ_SPAN_FROM_STR(" }"));
+  temp_span = az_span_copy_u8(temp_span, '\0');
+
+  return (char*)telemetry_payload;
+}
+
+static void send_telemetry()
+{
+  digitalWrite(LED_PIN, HIGH);
+  Serial.print(millis());
+  Serial.print(" Realtek Ameba-D Sending telemetry . . . ");
+  if (az_result_failed(az_iot_hub_client_telemetry_get_publish_topic(
+          &hub_client, NULL, telemetry_topic, sizeof(telemetry_topic), NULL)))
+  {
+    Serial.println("Failed az_iot_hub_client_telemetry_get_publish_topic");
+    return;
+  }
+
+  mqtt_client.publish(telemetry_topic, get_telemetry_payload(), false);
```

Comment:
can this api fail?

Reply:
It could, but unlikely. We are keeping it simple for now.